src/invidious/helpers/youtube_api.cr

```diff
@@ -3,22 +3,32 @@
 #
 
 # Hard-coded constants required by the API
-HARDCODED_API_KEY     = "AIzaSyAO_FJ2SlqU8Q4STEHLGCilw_Y9_11qcW8"
-HARDCODED_CLIENT_VERS = "2.20210330.08.00"
+HARDCODED_ANDROID_API_KEY     = "AIzaSyA8eiZmM1FaDVjRy-df2KTyQ_vz_yYM39w"
+HARDCODED_ANDROID_CLIENT_VERS = "16.20.35"
+HARDCODED_WEB_API_KEY         = "AIzaSyAO_FJ2SlqU8Q4STEHLGCilw_Y9_11qcW8"
+HARDCODED_WEB_CLIENT_VERS     = "2.20210623.03.00"
 
 ####################################################################
 # make_youtube_api_context(region)
 #
 # Return, as a Hash, the "context" data required to request the
 # youtube API endpoints.
 #
-def make_youtube_api_context(region : String | Nil) : Hash
+def make_youtube_api_context(region : String | Nil, client_name : String = "WEB") : Hash
+  case client_name
+  when "ANDROID"
+    hardcoded_android_client_vers = HARDCODED_ANDROID_CLIENT_VERS
+  when "WEB"
+    hardcoded_android_client_vers = HARDCODED_WEB_CLIENT_VERS
+  else
+    hardcoded_android_client_vers = HARDCODED_WEB_CLIENT_VERS
+  end
```

Comment:
Please use an enumeration instead of strings, to avoid heavy string comparisons (especially since we may add other clients in the future). And please also rename the intermediary variable (`hardcoded_android_client_vers`) to be more generic!

Reply:
How would you implement that with enum? I'm really not familiar with Crystal and I tried to use them, but I fail to understand how use them properly.
Here is what I tried:
``` crystal
# Hard-coded constants required by the API
enum HARDCODED_API_KEY : String
  ANDROID = "AIzaSyA8eiZmM1FaDVjRy-df2KTyQ_vz_yYM39w"
  WEB = "AIzaSyAO_FJ2SlqU8Q4STEHLGCilw_Y9_11qcW8"
end

enum HARDCODED_CLIENT_VERS : String
  ANDROID = "16.20.35"
  WEB = "2.20210623.03.00"
end

####################################################################
# make_youtube_api_context(region)
#
# Return, as a Hash, the "context" data required to request the
# youtube API endpoints.
#
def make_youtube_api_context(region : String | Nil, client_name : String = "WEB") : Hash
  case client_name
  when "ANDROID"
    hardcoded_client_vers = HARDCODED_API_KEY::ANDROID.value
  when "WEB"
    hardcoded_client_vers = HARDCODED_API_KEY::WEB.value
  else
    hardcoded_client_vers = HARDCODED_API_KEY::WEB.value
  end
  return {
    "client" => {
      "hl"            => "en",
      "gl"            => region || "US",       # Can't be empty!
      "clientName"    => client_name || "WEB", # Can't be empty!
      "clientVersion" => hardcoded_client_vers,
    },
  }
end
```

Reply:
See how it is done in the logging helper?
https://github.com/iv-org/invidious/blob/edb37c95e1c9dbf1abf2647fcfcd1453c38124c1/src/invidious/helpers/logger.cr#L1-L10
https://github.com/iv-org/invidious/blob/edb37c95e1c9dbf1abf2647fcfcd1453c38124c1/src/invidious/helpers/logger.cr#L39-L41

That would be kinda similar:
```crystal
enum ApiClients
  Web     = 0
  Android = 1
  SmartTv = 2 #maybe?
end
```

And then, for the function that will use them:
```crystal
def make_youtube_api_context(region : String | Nil, client_type = ApiClient::Web) : Hash
  case client_type
  when ApiClient::Web
    client_key  = HARDCODED_WEB_API_KEY
    client_vers = HARDCODED_WEB_CLIENT_VERSION
  when ApiClient::Android
    client_key  = HARDCODED_ANDROID_API_KEY
    client_vers = HARDCODED_ANDROID_CLIENT_VERSION
  when ApiClient::SmartTv
    client_key  = HARDCODED_TV_API_KEY
    client_vers = HARDCODED_TV_CLIENT_VERSION
  else
    # error
  end

  # Rest of code
end
```

Reply:
basically: enums (a.k.a enumerates) are a way in code to map human-understandable identifiers to numbers. This allows us to keep long names that are understandable by humans, without relying on expensive strings and string comparisons.

As we now have integers, we can directly use an array instead of a switch/case:
```crystal
HARDCODED_CLIENTS = [
  {key: "abcdef", name: "WEB", version: "1.2.3" },
  {key: "bcdefg", name: "MOBILE", version: "1.2.3" },
  {key: "cdefgh", name: "TV", version: "1.2.3" },
]
```
```crystal
# Access element
client_infos = HARDCODED_CLIENTS[client_type]
key = client_infos.key
version = client_infos.version
name = client_infos.name
```

We can also use a Hash, which is acessed the same way as above:
```crystal
HARDCODED_CLIENTS = [
  ApiClients::Web     => {key: "abcdef", name: "WEB", version: "1.2.3" },
  ApiClients::Android => {key: "bcdefg", name: "MOBILE", version: "1.2.3" },
  ApiClients::SmartTv => {key: "cdefgh", name: "TV", version: "1.2.3" },
]
```

Reply:
But we have to include the enum code in every single file that use a different client name for a function of youtube_api, isn't that a bit annoying?